File: src/nmath/vector.zig
const std = @import("std");
const expect = std.testing.expect;
const builtin = @import("builtin");
const math = std.math;

const constants = @import("constants.zig");
/// A 4D Vector used for points and directions and colors.
pub const Vec4 = @Vector(4, f32);
pub const Point = Vec4;
pub const Vector = Vec4;

//pub const Vec3 = @Vector(3, f32);

pub fn x(v: Vec4) f32 {
    return v[0];
}
pub fn y(v: Vec4) f32 {
    return v[1];
}
pub fn z(v: Vec4) f32 {
    return v[2];
}
pub fn w(v: Vec4) f32 {
    return v[3];
}

pub fn point(_x: f32, _y: f32, _z: f32) Vec4 {
    return .{ _x, _y, _z, 1 };
}

pub fn vector(_x: f32, _y: f32, _z: f32) Vec4 {
    return .{ _x, _y, _z, 0 };
}

pub fn isPoint(v: Vec4) bool {
    return v[3] == 1;
}

pub fn isVector(v: Vec4) bool {
    return v[3] == 0;
}

pub fn vec3(v: Vec4) @Vector(3, f32) {
    return .{ v[0], v[1], v[2] };
}

pub const FmtVec4 = std.fmt.Alt(Vec4, format);
fn format(v: Vec4, wr: *std.Io.Writer) !void {
    try wr.print("{d} {d} {d} {d}", .{ v[0], v[1], v[2], v[3] });
}
pub fn fmt(v: Vec4) FmtVec4 {
    return .{
        .data = v,
    };
}

pub fn approxEq(lhs: Vec4, rhs: Vec4) bool {
    return @reduce(.And, @abs(lhs - rhs) < splat(constants.epsilon));
}

pub fn approxZero(v: Vec4) bool {
    return @reduce(.And, @abs(v) < splat(constants.epsilon));
}

pub fn magnitude(v: Vec4) f32 {
    return @sqrt(magnitudeSquared(v));
}

pub fn magnitudeSquared(v: Vec4) f32 {
    return @reduce(.Add, v * v);
}

pub fn normalize(v: Vec4) Vec4 {
    std.debug.assert(!std.math.approxEqAbs(f32, magnitude(v), 0, constants.epsilon));

    const mag: Vec4 = @splat(magnitude(v));
    return v / mag;
}

/// We should only be using dot and cross product on vectors and not points.
pub fn dot(lhs: Vec4, rhs: Vec4) f32 {
    return @reduce(.Add, lhs * rhs);
}

/// 3D cross product, w is dropped
pub fn cross(lhs: Vec4, rhs: Vec4) Vec4 {
    return .{
        lhs[1] * rhs[2] - lhs[2] * rhs[1],
        lhs[2] * rhs[0] - lhs[0] * rhs[2],
        lhs[0] * rhs[1] - lhs[1] * rhs[0],
        0,
    };
}

pub inline fn splat(s: f32) Vec4 {
    return @as(Vec4, @splat(s));
}

pub fn reflect(in: Vector, normal: Vector) Vector {
    return in - normal * splat(2) * splat(dot(in, normal));
}

test "vector:reflect works at 45 degrees" {
    const v = vector(1, -1, 0);
    const n = vector(0, 1, 0);
    const r = reflect(v, n);
    const expected = vector(1, 1, 0);
    try expect(approxEq(r, expected));
}

test "vector:reflect works at slanted angles" {
    const v = vector(0, -1, 0);
    const root_2_over_2 = @sqrt(2.0) / 2.0;
    const n = vector(root_2_over_2, root_2_over_2, 0);
    const r = reflect(v, n);
    const expected = vector(1, 0, 0);
    try expect(approxEq(r, expected));
}

File: pkg/nmath/vector.go
package nmath

import "math"

type Vec4 struct {
	X, Y, Z, W float64
}

func NewVec4(x, y, z, w float64) Vec4 {
	return Vec4{x, y, z, w}
}

func NewPoint4(x, y, z float64) Vec4 {
	return Vec4{x, y, z, 1}
}
func NewVector4(x, y, z float64) Vec4 {
	return Vec4{x, y, z, 0}
}

func (v Vec4) DropW() Vec3 {
	return Vec3{v.X, v.Y, v.Z}
}

func (lhs Vec4) ApproxEq(rhs Vec4) bool {
	return ApproxEq(lhs.X, rhs.X) &&
		ApproxEq(lhs.Y, rhs.Y) &&
		ApproxEq(lhs.Z, rhs.Z) &&
		ApproxEq(lhs.W, rhs.W)
}

func (lhs Vec4) Add(rhs Vec4) Vec4 {
	return Vec4{
		lhs.X + rhs.X,
		lhs.Y + rhs.Y,
		lhs.Z + rhs.Z,
		lhs.W + rhs.W,
	}
}

func (lhs Vec4) Sub(rhs Vec4) Vec4 {
	return Vec4{
		lhs.X - rhs.X,
		lhs.Y - rhs.Y,
		lhs.Z - rhs.Z,
		lhs.W - rhs.W,
	}
}

func (lhs Vec4) Mult(rhs float64) Vec4 {
	return Vec4{
		lhs.X * rhs,
		lhs.Y * rhs,
		lhs.Z * rhs,
		lhs.W * rhs,
	}
}

func (lhs Vec4) Div(rhs float64) Vec4 {
	return Vec4{
		lhs.X / rhs,
		lhs.Y / rhs,
		lhs.Z / rhs,
		lhs.W / rhs,
	}
}

func (lhs Vec4) Dot(rhs Vec4) float64 {
	return lhs.X*rhs.X + lhs.Y*rhs.Y + lhs.Z*rhs.Z + lhs.W*rhs.W
}

// 3D cross product, w is dropped
func (lhs Vec4) Cross(rhs Vec4) Vec4 {
	return Vec4{
		lhs.Y*rhs.Z - lhs.Z*rhs.Y,
		lhs.Z*rhs.X - lhs.X*rhs.Z,
		lhs.X*rhs.Y - lhs.Y*rhs.X,
		0,
	}
}

// Magnitude squared
func (v Vec4) Mag2() float64 {
	return v.Dot(v)
}

func (v Vec4) Mag() float64 {
	return math.Sqrt(v.Mag2())
}

func (v Vec4) Normalize() Vec4 {
	mag := v.Mag()
	return Vec4{
		v.X / mag,
		v.Y / mag,
		v.Z / mag,
		v.W / mag,
	}
}

// -----------------------------------------------------------------------
// Vec3
// -----------------------------------------------------------------------

type Vec3 struct {
	X, Y, Z float64
}

func NewVec3(x, y, z float64) Vec3 {
	return Vec3{x, y, z}
}

func (lhs Vec3) ApproxEq(rhs Vec3) bool {
	return ApproxEq(lhs.X, rhs.X) &&
		ApproxEq(lhs.Y, rhs.Y) &&
		ApproxEq(lhs.Z, rhs.Z)
}

func (v Vec3) AsPoint4() Vec4 {
	return Vec4{v.X, v.Y, v.Z, 1}
}

func (v Vec3) AsVector4() Vec4 {
	return Vec4{v.X, v.Y, v.Z, 0}
}

func (v Vec3) WithW(w float64) Vec4 {
	return Vec4{v.X, v.Y, v.Z, w}
}

func (v Vec3) Neg() Vec3 {
	return Vec3{-v.X, -v.Y, -v.Z}
}

func (lhs Vec3) Add(rhs Vec3) Vec3 {
	return Vec3{lhs.X + rhs.X, lhs.Y + rhs.Y, lhs.Z + rhs.Z}
}

func (lhs Vec3) Sub(rhs Vec3) Vec3 {
	return Vec3{lhs.X - rhs.X, lhs.Y - rhs.Y, lhs.Z - rhs.Z}
}

func (lhs Vec3) Mult(rhs float64) Vec3 {
	return Vec3{
		lhs.X * rhs,
		lhs.Y * rhs,
		lhs.Z * rhs,
	}
}

func (lhs Vec3) Div(rhs float64) Vec3 {
	return Vec3{
		lhs.X / rhs,
		lhs.Y / rhs,
		lhs.Z / rhs,
	}
}

func (lhs Vec3) Dot(rhs Vec3) float64 {
	return lhs.X*rhs.X + lhs.Y*rhs.Y + lhs.Z*rhs.Z
}

func (lhs Vec3) Cross(rhs Vec3) Vec3 {
	return Vec3{
		lhs.Y*rhs.Z - lhs.Z*rhs.Y,
		lhs.Z*rhs.X - lhs.X*rhs.Z,
		lhs.X*rhs.Y - lhs.Y*rhs.X,
	}
}

func (in Vec3) Reflect(normal Vec3) Vec3 {
	return in.Sub(normal.Mult(2).Mult(in.Dot(normal)))
}

func (v Vec3) Mag2() float64 {
	return v.Dot(v)
}

func (v Vec3) Mag() float64 {
	return math.Sqrt(v.Mag2())
}

func (v Vec3) Normalize() Vec3 {
	m := v.Mag()
	if m == 0 {
		return v
	}
	return Vec3{v.X / m, v.Y / m, v.Z / m}
}

func (v Vec3) AsColor() Color {
	return Color{v.X, v.Y, v.Z, 1}
}

func (from Vec3) LookAt(to, up Vec3) Mat4 {
	forward := to.Sub(from).Normalize()
	left := forward.Cross(up.Normalize())
	true_up := left.Cross(forward)
	orientation := Mat4{
		left.X, left.Y, left.Z, 0,
		true_up.X, true_up.Y, true_up.Z, 0,
		-forward.X, -forward.Y, -forward.Z, 0,
		0, 0, 0, 1,
	}

	return orientation.Translate(-from.X, -from.Y, -from.Z)
}
